includes/panopto_derivatives.inc

```diff
@@ -5,37 +5,384 @@
+    if ($session) {
+      $upload_id = $session['ID'];
+      $upload_target = $session['UploadTarget'];
+      $manifest_file = islandora_video_streaming_panopto_create_manifest_for_video($filePath);
+      if (isset($manifest_file)) {
+        try {
+          islandora_video_streaming_panopto_upload_file($upload_target, $manifest_file);
+          islandora_video_streaming_panopto_upload_file($upload_target, $filePath);
+        } catch (Exception $ex) {
+          $messages = array(
+            'message' => t('Panopto upload error: @exception'),
+            'message_sub' => array('@exception' => $ex->getMessage()),
+            'type' => 'watchdog',
+            'severity' => WATCHDOG_ERROR,
+          );
+          $failureMessage['messages'][] = $messages;
+          watchdog_exception('islandora_video_streaming_panopto', $ex);
+        }
+        $response = islandora_video_streaming_panopto_finish_upload($uploadendpoint, $session, $token);
+        if (isset($response)) {
+          try {
+            $success = islandora_video_streaming_monitor_progress($uploadendpoint, $upload_id, $token);
+          } catch (Exception $ex) {
+            $messages = array(
+            'message' => t('Panopto progress error: @exception'),
+            'message_sub' => array('@exception' => $ex->getMessage()),
+            'type' => 'watchdog',
+            'severity' => WATCHDOG_ERROR,
+            );
+            $failureMessage['messages'][] = $messages;
+            watchdog_exception('islandora_video_streaming_panopto', $ex);
+            return $failureMessage;
+          }
+          $streaming['upload_id'] = $session['ID'];
+        } else {
+          $messages = array(
+            'message' => t('Issue with finishing upload for object: @pid, response: @response'),
+            'message_sub' => array(
+              '@pid' => $object->id,
+              '@response' => $response,
+            ),
+            'type' => 'watchdog',
+            'severity' => WATCHDOG_ERROR,
+          );
+          $failureMessage['messages'][] = $messages;
+        }
+      } 
+    } else {
+      $messages = array(
+        'message' => t('Failed to retrieve session for object: @pid'),
         'message_sub' => array('@pid' => $object->id),
         'type' => 'watchdog',
         'severity' => WATCHDOG_ERROR,
-      ),
-    );
-
-    //TODO: upload, record result in $streaming
- 
+      );
+      $failureMessage['messages'][] = $messages;
+    }
+    file_delete($archival_file['file'], true);
     if ($streaming) {
        // Write the data to a temporary file.
       $temp_name = drupal_tempnam('temporary://', 'file');
       if (file_put_contents($temp_name, json_encode($streaming)) === FALSE) {
-        $failureMessage['messages'][0]['message'] = t('Failed to create temporary file @filename for object @pid when uploading to Panopto.');
-        $failureMessage['messages'][0]['message_sub']['@filename'] = $temp_name;
+        $messages = array(
```

Comment:
This is a great example of a well-constructed error message.  It specifies the object being worked on, and the operation with failed, with the filename where the attempted write failed.  The only way to improve it would be to report on the actual error message from `file_put_contents`, but `file_put_contents` doesn't give us one.